parlai/messenger/core/messenger_manager.py

```diff
@@ -422,23 +429,27 @@ def setup_server(self):
             should_print=True
         )
 
-    def setup_socket(self):
-        """Set up socket to start communicating to workers"""
-        shared_utils.print_and_log(logging.INFO,
-                                   'Local: Setting up WebSocket...',
-                                   should_print=True)
-        self.app_token = None
-        if self.opt.get('force_page_token'):
-            pass
-        else:
+    # override if permission needed externally
+    def get_app_token(self):
+        token = None
+        if not self.opt.get('force_page_token'):
             if not os.path.exists(os.path.expanduser('~/.parlai/')):
                 os.makedirs(os.path.expanduser('~/.parlai/'))
             access_token_file_path = '~/.parlai/messenger_token'
             expanded_file_path = os.path.expanduser(access_token_file_path)
             if os.path.exists(expanded_file_path):
                 with open(expanded_file_path, 'r') as access_token_file:
-                    self.app_token = access_token_file.read()
+                    token = access_token_file.read()
+        return token
 
+    def setup_socket(self):
+        """Set up socket to start communicating to workers"""
+        if self.need_server_setup:
+            shared_utils.print_and_log(logging.INFO,
+                                       'Local: Setting up WebSocket...',
+                                       should_print=True)
+
+        self.app_token = self.get_app_token()
         # cache the app token
         if self.app_token is None:
             self.app_token = input(
```

Comment:
This logic can also go into the `self.get_app_token()` call, as it is still part of the flow of getting an app token.